### ISSUE TITLE
Additional /back teleport points

### ISSUE THREAD

**Issue by seema84** [EssentialsX/Essentials#1720]
can you please add more than one back points?
maybe the last 10 how it does commandbook

**Comment by mishyy** [comment]
This sounds like an interesting feature, any suggestions on how to implement it?

**Comment by seema84** [comment]
it works as one back point, only the latest points need to cache

**Comment by drtshock** [comment]
can you go into more detail? :)

**Comment by seema84** [comment]
example:
1. use /back and go to the last point (tp,death etc)
2. use /back and go to the second last point (tp,death etc)
3. use /back and go to the third last point (tp,death etc)
4. etc

**Comment by TNTUP** [comment]
That seems to be a great idea but players who uses /back at Point A to B like 12 times would be confusing or it would cache?^^

**Comment by seema84** [comment]
yes i would cache even if it seems pointless

**Comment by pop4959** [comment]
Sounds useful, although I would suggest by default keeping the normal behavior (where it takes you back and forth between point A and B). Could use something like `/back 2` to go back two teleportations, similar to how plugins like WorldEdit have `//undo 2`, etc.

**Comment by werogg** [comment]
Hello, I'm currently working on this feature but have some doubts regarding it, we will be actually caching the last locations of the user, so I expect the idea is not persisting any data (not like last location that is currently serialized and persisted).

So I have some questions about the possible implementations:

- Should we just cache this data in memory while the user is connected and reset it every time the user reconnects.
- We persist the data as lastlocation does but we set a limit of back locations and we delete in a FIFO. In this implementation we could also set different limits for different permissions, so a user can have 5 backs while another can have 20.

**Comment by JRoy** [comment]
> Hello, I'm currently working on this feature but have some doubts regarding it, we will be actually caching the last locations of the user, so I expect the idea is not persisting any data (not like last location that is currently serialized and persisted).
> 
> So I have some questions about the possible implementations:
> 
> * Should we just cache this data in memory while the user is connected and reset it every time the user reconnects.
> * We persist the data as lastlocation does but we set a limit of back locations and we delete in a FIFO. In this implementation we could also set different limits for different permissions, so a user can have 5 backs while another can have 20.

The former

**Comment by werogg** [comment]
So just caching the data true? (Not sure about "the former" meaning in this context)

**Comment by mdcfe** [comment]
I'd say just cache them in memory. You can still store them in a limited size queue.

It wouldn't be too difficult to persist these locations to disk, but I also don't really see how useful this would be to players.

**Comment by werogg** [comment]
Actually the lastlocation is fully persisted, so even if the user disconnects or the server is restarted, he is allowed to use /back again. So this option would be unavailable in the `/back <integer>` feature.

**Comment by mdcfe** [comment]
The implementation of this wouldn't be difficult, but I'm not sure I get how and why people would actually use this feature. 

How useful is it to be able to go back to your 2nd/3rd/4th last teleport location? You might remember the last two or three times you teleported or died but I can't imagine most players would remember much more than that.

You could give users some way to check what the coordinates of their last N locations actually are (eg `/backinfo 3`), but even if they could see the coordinates and world, there's no guarantee that these would mean anything to the average player. If they can't remember where the locations are, they will either spam `/back` to see each location or they'll just not bother using the queue at all.

The other problem is what do you do if someone repeatedly `/back`s? Their queue will fill up with duplicate locations (or near duplicates), making anything beyond the last few useless for them. It would also be a pain to mentally track those locations at the actual numbers, and if you accidentally `/back` to the wrong one you might just lose one of the locations.

Both of these problems could be solved by letting players pin `/back` locations and name them... at which point you've just reinvented `/home`.

I guess there's some argument for being able to cycle back to an older location just to create a home there? I'm just concerned that the practical use cases for this in its current form are quite limited, and I think I need to see some actual scenarios where this might be more useful to players and server owners than simply setting homes.

**Comment by seema84** [comment]
Why should you not be able to remember? In fact, it happens to "me" quite often that I would like to quickly return two or three points after a teleport. 

Since it's been many years now, I don't really remember how commandbock implemented it.

**Comment by pop4959** [comment]
I agree with @mdcfe on this behavior most likely leading to a lot of confusion. Perhaps there is a better way than incorporating this directly into back. Maybe a `/recent` command which lists your last X recent unique locations? Then with `/back #` instead of going that many locations back, it just goes to the location listed in your recent. Or maybe something else. But as-is this is way too confusing to expect most players to find useful.

**Comment by seema84** [comment]
You are probably right that a reverse command without addition and as default could cause confusion. 
Your alternative approaches don't sound bad in principle.

**Comment by merenze** [comment]
> Maybe a `/recent` command which lists your last X recent unique locations?

Lists them how? Not that I disagree, just that I wouldn't know how to implement this; I'm not sure (world, x, y, z) is meaningful enough to be useful to most players most of the time.

**Comment by JRoy** [comment]
> > Maybe a `/recent` command which lists your last X recent unique locations?
> 
> Lists them how? Not that I disagree, just that I wouldn't know how to implement this; I'm not sure (world, x, y, z) is meaningful enough to be useful to most players most of the time.

I would say maybe listen them as to what caused the teleport. So like "/tpa to Player1" or "/warp spawn".

**Comment by maxlmb2023** [comment]
I think you can delete the same point in a teleport. e.g., the back list is `ABC` and I use `/back` to go to `B`, so the list changes to `ACB` instead `ABCB`.
